### ISSUE TITLE
 Error processing tar file(exit status 1): Error cleaning up after pivot: remove /.pivot_root767627221: device or resource busy.

### ISSUE THREAD

**Issue by Xelawk** [moby/moby#40008]
**Description**

**Steps to reproduce the issue:**
1. docker run hello-world

**Describe the results you received:**
~# docker run hello-world
Unable to find image 'hello-world:latest' locally
latest: Pulling from library/hello-world
1b930d010525: Extracting     977B/977B
docker: failed to register layer: Error processing tar file(exit status 1): Error cleaning up after pivot: remove /.pivot_root767627221: device or resource busy.

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           19.03.2
 API version:       1.40
 Go version:        go1.12.8
 Git commit:        6a30dfc
 Built:             Thu Aug 29 05:28:19 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.2
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.8
  Git commit:       6a30dfc
  Built:            Thu Aug 29 05:26:54 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**

```
Client:
 Debug Mode: false

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 0
 Server Version: 19.03.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc version: 425e105d5a03fabd737a126ad93d62a9eeede87f
 init version: fec3683
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 4.4.0-146-generic
 Operating System: Ubuntu 16.04.6 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 28
 Total Memory: 110.2GiB
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
hm, I don't recall having seen this error before. Did you keep the daemon and system logs from around the time this happened? Do you have software (e.g. anti-virus, or monitoring software) running on the host that could be keeping these files opened?

@kolyshkin any ideas?

**Comment by kolyshkin** [comment]
I haven't seen that either. The error comes from here:

https://github.com/moby/moby/blob/bb8438bb002f05310e21166ef39f047efe3eb795/pkg/chrootarchive/chroot_linux.go#L67

and this code was last changed in 2017. I have no idea why this could ever happen.

@Xelawk As suggested earlier, it would be beneficial to see the system logs around that time.

**Comment by Envek** [comment]
Faced the same problem during `docker pull` command.

```
$ docker pull docker.pkg.github.com/some/private/image:latest
latest: Pulling from some/private/image
df20fa9351a1: Already exists 
7da614efc9d6: Already exists 
36965dd2685b: Already exists 
8db9b749023b: Already exists 
a22ef58fc35c: Extracting [==================================================>]  17.18kB/17.18kB
b6a0bca67fbf: Download complete 
4731bb517d3b: Download complete 
656444004f2d: Download complete 
7ddac86bb035: Download complete 
failed to register layer: Error processing tar file(exit status 1): Error cleaning up after pivot: remove /.pivot_root277226559: device or resource busy
```
Next attempts failing immediately (with any image, even public one):
```
$ docker pull docker.pkg.github.com/some/private/image:latest
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```
The only info I can found in `journalctl` is:
```
Aug 19 12:22:43 Envek-NUC dockerd[1353]: time="2020-08-19T12:22:43.526767832+03:00" level=info msg="Attempting next endpoint for pull after error: failed to register layer: Error processing tar file(exit status 1): Error cleaning up after pivot: remove /.pivot_root277226559: device or resource busy"
```

`systemctl` says that docker is working:
<details>
<summary>`sudo systemctl status docker`</summary>

```
$ sudo systemctl status docker                                              
● docker.service - Docker Application Container Engine
     Loaded: loaded (/usr/lib/systemd/system/docker.service; enabled; vendor preset: disabled)
     Active: active (running) since Wed 2020-08-19 12:21:42 MSK; 17min ago
TriggeredBy: ● docker.socket
       Docs: https://docs.docker.com
   Main PID: 1353 (dockerd)
      Tasks: 30
     Memory: 227.6M
     CGroup: /system.slice/docker.service
             └─1353 /usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock

Aug 19 12:21:41 Envek-NUC dockerd[1353]: time="2020-08-19T12:21:41.847801513+03:00" level=warning msg="Your kernel does not support cgroup rt runtime"
Aug 19 12:21:41 Envek-NUC dockerd[1353]: time="2020-08-19T12:21:41.847904990+03:00" level=info msg="Loading containers: start."
Aug 19 12:21:42 Envek-NUC dockerd[1353]: time="2020-08-19T12:21:42.073824102+03:00" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set >
Aug 19 12:21:42 Envek-NUC dockerd[1353]: time="2020-08-19T12:21:42.105989916+03:00" level=info msg="Loading containers: done."
Aug 19 12:21:42 Envek-NUC dockerd[1353]: time="2020-08-19T12:21:42.252641747+03:00" level=warning msg="Not using native diff for overlay2, this may cause degraded performance for building images: kernel has CON>
Aug 19 12:21:42 Envek-NUC dockerd[1353]: time="2020-08-19T12:21:42.252901832+03:00" level=info msg="Docker daemon" commit=48a66213fe graphdriver(s)=overlay2 version=19.03.12-ce
Aug 19 12:21:42 Envek-NUC dockerd[1353]: time="2020-08-19T12:21:42.253573843+03:00" level=info msg="Daemon has completed initialization"
Aug 19 12:21:42 Envek-NUC dockerd[1353]: time="2020-08-19T12:21:42.282312388+03:00" level=info msg="API listen on /run/docker.sock"
Aug 19 12:21:42 Envek-NUC systemd[1]: Started Docker Application Container Engine.
Aug 19 12:22:43 Envek-NUC dockerd[1353]: time="2020-08-19T12:22:43.526767832+03:00" level=info msg="Attempting next endpoint for pull after error: failed to register layer: Error processing tar file(exit status>
```
</details>

Restarting docker service doesn't help. Only system reboot helps.

OS: Manjaro Linux. Docker installed via package manager.

**Comment by thaJeztah** [comment]
@Envek that looks like a custom build of docker (perhaps it's a distro package?) I see its built with Golang 1.14, which has breaking changes wrt handling of syscalls, which could potentially play a role. Are you seeing the same problem with packages from download.docker.com ?

**Comment by Envek** [comment]
> perhaps it's a distro package?

Yes, it is a distro package [docker 1:19.03.12-2](https://www.archlinux.org/packages/community/x86_64/docker/) from Arch Linux. And it worked well for a year and broke just recently. Maybe after last update.

There is 2 years old bug in Arch packages [FS#58690](https://bugs.archlinux.org/task/58690) but I can't find any clues in it.

Any operation with docker fails and then system starts to behave weirdly.

**Comment by thaJeztah** [comment]
They definitely shouldn't be building with Go 1.14, but I've seen more issues with the Arch packages (there was one earlier this week where they used incorrect compile options for containerd, causing things to break in subtle ways).

Might be good to see if it reproduces on the official docker packages, or otherwise report to the Arch linux package issue tracker

**Comment by Envek** [comment]
Linux Kernel upgrade solved it for me: upgraded from 4.19.138 to 5.4.57

**Comment by sirex** [comment]
Yes upgrading kernel worked for me too. Just a quick reminder how to upgrade a kernel:

```
sudo pacman -S linux54
```

Or you can do it via GUI: https://wiki.manjaro.org/index.php/Manjaro_Kernels

**Comment by thaJeztah** [comment]
Let me go ahead and close this issue, but feel free to continue the conversation

**Comment by joseph-r-hamilton** [comment]
I have the exact same problem, although I trigger it in a different way.  I use packer to create containers.  I ran into this error when copying a file to the container.  But after the "docker: Upload failed" text, everything is the same.  The odd message about pivot.  The truly bizarre behavior of the symlinks failing.  Etc. etc.

I experimented a bit with downgrading and rolling back to a snapshot and upgrading specific things.

This odd problem behavior isn't solely due to the change in the docker level.

I tested the permutations of:

docker: 19.03.11-1 and 19.03.12-2
containerd: 1.3.4-2 and 1.4.0-2
linux: 4.19.126-1 and 4.19.143-1

The definitive factor was the change in the linux kernel.  On 4.19.126, I had no issue no matter the versions of the other things.  On 4.19.143, this problem case arose no matter the versions of the other things.  I first experienced the problem on 4.19.141.  So somewhere between 126 and 141, something changed in the kernel that created this problem.

This, to me, dovetails with others' experience that upgrading the kernel to 5 resolved the issue.

**Comment by Minilfat** [comment]
I faced exactly the same issue on Manjaro with 4.19.143-1-MANJARO x86_64 after docker updated to 19.03.12-ce
Upgrading kernel to 5.4.62-1-MANJARO x86_64 solved problem.

**Comment by FlorianKnapp** [comment]
I got the same issue 2 hours ago after i worked already more than 4 hours today with docker and weeks before. What killed it in the end was doing ```docker-compose down --rmi=all```. After that docker won't work anymore... Additionally my system was getting into a weird state where everything with sudo involved is taking about 10 seconds to execute. I'll try the kernel suggestion :-P

**Comment by Envek** [comment]
> Additionally my system was getting into a weird state where everything with sudo involved is taking about 10 seconds to execute.

Yes, I also noticed this.

**Comment by yssource** [comment]
The same problem happened to me, also.

And `/etc/mtab` is broken.

**Comment by yssource** [comment]
`uname -a`

`Linux jimmy-arch 4.19.152-1-MANJARO #1 SMP Sat Oct 17 10:51:27 UTC 2020 x86_64 GNU/Linux
`

**Comment by Imeda** [comment]
Error still remains. Cannot update to Kernel 5.4 as i get the following error:

Packages (1) linux54-5.4.74-1

Total Download Size:    85.02 MiB
Total Installed Size:  140.16 MiB

:: Proceed with installation? [Y/n] Y
error: could not open file: /etc/mtab: No such file or directory
error: could not determine filesystem mount points
error: failed to commit transaction (unexpected error)
Errors occurred, no packages were upgraded.

**Comment by ilmagico** [comment]
I just had this same problem, same OS, Manjaro, with kernel 4.19.152-1-MANJARO. Unfortunately, I cannot upgrade to 5.4, or anything higher, because my laptop randomly freezes with any of those kernels.

Did anybody find out the root cause, or should I try to downgrade my kernel to 4.19.126-1 (or, well, the latest 4.19 working kernel)?

**Comment by joseph-r-hamilton** [comment]
Bug was introduced into Manjaro linux49 (https://gitlab.manjaro.org/packages/core/linux419)
at commit 98da16f.

Specifically, the problem is **proc_mounts.patch** which was added then.

Last good release was 4.19.133 (if it was released - cannot be sure) or 4.19.132.

You can build the latest kernel without proc_mounts.patch.  Three modifications are required
in PKGBUILD: the call to patch; the source; the checksum.

I have tested this on 4.19.152.  It does fix the problem. 

Applying proc_mounts.patch to Manjaro 5.4 recreates the exact same problem. 

Applying proc_mounts.patch to ArchLinux 5.4 does not create the problem. 

I think the aufs patches are tripping over themselves.  But I'm not yet sure. 

Furthermore, I believe the problem would affect containers in general and not just docker. 

I have thoughts about the true root cause.  But they're not much further beyond speculation. 
So I decided to post this much at the moment to help anyone stuck.

**Comment by ilmagico** [comment]
@joseph-r-hamilton Thank you for the info! At the next available occasion, I'm going to first try the latest stable kernel (5.9 I believe) to see if this is fixed *and* finally the random freezes are fixed as well, if not, I'm going to rebuild 4.19.152 without `proc_mounts.patch` while waiting for an upstream fix.

**Comment by joseph-r-hamilton** [comment]
Oh...  yeah...

I forgot to mention, the proc_mounts.patch cannot be applied to 5.9.  The underlying files changed too much and I'm not aware of an updated version from the aufs folk.

**Comment by ilmagico** [comment]
> I forgot to mention, the proc_mounts.patch cannot be applied to 5.9.

Wasn't planning to :) Either 5.9/5.10 Manjaro unmodified, or 4.19.x with `proc_mounts.patch` removed. I'll update with the results.

**Comment by ilmagico** [comment]
So far, I updated to 5.9.8-2-MANJARO and it seems to work correctly, docker works (the problem in this issue doesn't happen), and it didn't randomly freeze in the last couple of days at least (finally! been waiting for years!), so I think I'm going to stick to it. If problems start to happen I'll go back to the patched 4.19 route instead. Thanks!
